CMakeLists.txt

```diff
@@ -0,0 +1,250 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2013-2022 The Foundry Visionmongers Ltd
+
+#-----------------------------------------------------------------------
+# CMake preamble.
+
+# FindPython's Development.Module component was added in 3.18, and
+# pybind11 recommends 3.18.2+.
+cmake_minimum_required(VERSION 3.18.2)
+
+# Additional include directories for CMake utils.
+list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/cmake)
+
+if (APPLE AND NOT DEFINED CMAKE_OSX_DEPLOYMENT_TARGET)
+    # The value of this variable should be set prior to the first
+    # project() command invocation because it may influence
+    # configuration of the toolchain and flags.
+    set(CMAKE_OSX_DEPLOYMENT_TARGET "10.13" CACHE STRING "Minimum OS X deployment version")
+endif ()
+
+# Forbid in-source builds.
+set(CMAKE_DISABLE_IN_SOURCE_BUILD ON)  # NB: currently undocumented.
```

Comment:
Didn't spot these two variables in the first pass.

This https://gitlab.kitware.com/cmake/cmake/-/issues/18403 says it's never going to be documented, and for reasons given in the notes on that link.

What was the reason that these are here? What are they trying to solve?

Reply:
Ah yeah, I did see that (hence the `NB`) but didn't think the arguments in there were particularly concerning. The reason to disallow in-source builds is just that it's bad practice. Other projects do similar, e.g.

* https://github.com/OpenImageIO/oiio/blob/65c457d6f5b48b10727ed9e9d0eaf311bd22f245/CMakeLists.txt#L72
* https://github.com/AcademySoftwareFoundation/OpenColorIO/blob/1d8a28c2e2f842de053cef25887b7d5bc89dc438/CMakeLists.txt#L58
* https://github.com/cpp-best-practices/project_options/blob/398a122a387cefb697a751c64fb84574a7061694/src/PreventInSourceBuilds.cmake#L4

To be fair, they all work it out, rather than using the (undocumented) variable(s). 

Do you think we should just remove this condition entirely, or opt for a hand-rolled check over these variables?

Reply:
I agree it is bad practice to do in-source builds, but I must admit to not having seen anyone explicitly disallow it before. It's interesting that it's "out there" in the wild.

I usually let people do what they want. But document the preferred workflow.

The reason being, if if you add restrictions, there's bound to be someone who wants to do it the other way, and then you get support questions.
On the other hand, allowing it, and only testing out-of-source builds, may also leave it open for support questions.

Seems to be a no-win.

I will admit to being concerned on using undocumented features (CMake could yank them).

Reply:
Removed use of the undocumented flag in favour of a hand-rolled check in 82d75097